File: platforms/ios/Packages/FunputKit/Sources/KeyboardRenderer/Emoji/EmojiCatalog.swift
import Foundation

public enum EmojiCategory: String, CaseIterable, Codable, Hashable, Sendable {
    case recent
    case smileysPeople = "smileys_people"
    case animalsNature = "animals_nature"
    case foodDrink = "food_drink"
    case activities
    case travelPlaces = "travel_places"
    case objects
    case symbols
    case flags

    public var accessibilityLabel: String {
        switch self {
        case .recent: "Dùng gần đây"
        case .smileysPeople: "Mặt cười và con người"
        case .animalsNature: "Động vật và thiên nhiên"
        case .foodDrink: "Đồ ăn và thức uống"
        case .activities: "Hoạt động"
        case .travelPlaces: "Du lịch và địa điểm"
        case .objects: "Đồ vật"
        case .symbols: "Biểu tượng"
        case .flags: "Cờ"
        }
    }
}

public struct EmojiItem: Codable, Hashable, Sendable {
    public let glyph: String
    public let name: String
    public let category: EmojiCategory

    public init(glyph: String, name: String, category: EmojiCategory) {
        self.glyph = glyph
        self.name = name
        self.category = category
    }
}

public struct EmojiCatalog: Codable, Sendable {
    public let version: String
    public let emojis: [EmojiItem]

    public init(version: String, emojis: [EmojiItem]) {
        self.version = version
        self.emojis = emojis
    }

    public func items(in category: EmojiCategory) -> [EmojiItem] {
        emojis.filter { $0.category == category }
    }
}

File: platforms/android/keyboard-ui/src/main/java/app/funput/funput/keyboard/ui/emoji/EmojiCatalog.kt
package app.funput.funput.keyboard.ui.emoji

enum class EmojiCategory(val wireName: String, val label: String) {
    RECENT("recent", "Gần đây"),
    SMILEYS_PEOPLE("smileys_people", "Mặt cười & con người"),
    ANIMALS_NATURE("animals_nature", "Động vật & thiên nhiên"),
    FOOD_DRINK("food_drink", "Đồ ăn & thức uống"),
    ACTIVITIES("activities", "Hoạt động"),
    TRAVEL_PLACES("travel_places", "Du lịch & địa điểm"),
    OBJECTS("objects", "Đồ vật"),
    SYMBOLS("symbols", "Ký hiệu"),
    FLAGS("flags", "Cờ");

    companion object {
        fun fromWireName(value: String): EmojiCategory? =
            entries.firstOrNull { it.wireName == value }
    }
}

data class EmojiItem(
    val glyph: String,
    val name: String,
    val localizedName: String?,
    val searchTerms: List<String>,
    val category: EmojiCategory,
    val variants: List<String> = emptyList(),
) {
    val accessibilityLabel: String get() = localizedName ?: name
}

data class EmojiCatalog(
    val version: String,
    val annotationVersion: String,
    val emojis: List<EmojiItem>,
) {
    fun items(category: EmojiCategory): List<EmojiItem> =
        emojis.filter { it.category == category }

    companion object {
        val Empty = EmojiCatalog("", "", emptyList())
    }
}
